src/Parallel/PhaseControl/CheckpointAndExitAfterWallclock.hpp

```diff
@@ -0,0 +1,295 @@
+          "only be altered by the phase change arbitration in the Main "
+          "chare, so no reduction data should be provided.");
+    }
+  };
+  using main_combine_method = combine_method;
+};
+
+/// Stores whether the checkpoint and exit has been requested.
+///
+/// Combinations are performed via `funcl::Or`, as the phase in question should
+/// be chosen if any component requests the jump.
+struct CheckpointAndExitRequested {
+  using type = bool;
+
+  using combine_method = funcl::Or<>;
+  using main_combine_method = funcl::Or<>;
+};
+
+}  // namespace Tags
+
+/*!
+ * \brief Phase control object that runs the WriteCheckpoint and Exit phases
+ * after a specified amount of wallclock time has elapsed.
+ *
+ * This phase control is useful for running SpECTRE executables performing
+ * lengthy computations that may exceed a supercomputer's wallclock limits.
+ * Writing a single checkpoint at the end of the job's allocated time allows
+ * the computation to be continued, while minimizing the disc space taken up by
+ * checkpoint files.
+ *
+ * Note that this phase control is not a trigger on wallclock time. Rather,
+ * it checks the elapsed wallclock time when called, likely from a global sync
+ * point triggered by some other mechanism, e.g., at some slab boundary.
+ * Therefore, the WriteCheckpoint and Exit phases will run the first time
+ * this phase control is called after the specified wallclock time has been
+ * reached.
+ *
+ * \warning the global sync points _must_ be triggered often enough to ensure
+ * there will be at least one sync point (i.e., one call to this phase control)
+ * in the window between the requested checkpoint-and-exit time and the time at
+ * which the batch system will kill the executable. To make this more concrete,
+ * consider this example: when running on a 12-hour queue with a
+ * checkpoint-and-exit requested after 11.5 hours, there is a 0.5-hour window
+ * for a global sync to occur, the checkpoint files to be written to disc, and
+ * the executable to clean up. In this case, triggering a global sync every
+ * 2-10 minutes might be desirable. Matching the global sync frequency with the
+ * time window for checkpoint and exit is the responsibility of the user!
+ */
+template <typename Metavariables,
+          typename PhaseChangeRegistrars = tmpl::list<
+              Registrars::CheckpointAndExitAfterWallclock<Metavariables>>>
+struct CheckpointAndExitAfterWallclock
+    : public PhaseChange<PhaseChangeRegistrars> {
+  // This PhaseChange only makes sense if Metavars has a WriteCheckpoint phase
+  static_assert(Parallel::Algorithm_detail::has_WriteCheckpoint_v<
+                    typename Metavariables::Phase>,
+                "Requested to write checkpoints but Metavariables::Phase "
+                "doesn't have a WriteCheckpoint phase");
+
+  CheckpointAndExitAfterWallclock() noexcept {
+    time_at_start_ = std::chrono::system_clock::now();
```

Comment:
I can't find documentation for `CmiWallTime` (the function we wrap in our `sys::parallel::wall_time`), and so I'm not sure I agree with you that I can just call `wall_time()` and assume it started measuring from zero when I started my executable. I think the only safe thing to do is call `wall_time()` twice, which will miss startup costs just like calling the STL function twice (with the added displeasure that it's undocumented instead of documented). Are you aware of a documentation/guarantee on the behavior of `CmiWallTime` so we can rely on a particular behavior from `wall_time`?

Reply:
I think it always counts from the start, that's what I've observed in practice (and we assume that in a couple other places, like the time observer). I've asked the Charm++ folks for confirmation :)

Reply:
I'm not sure why we use `CmiWallTimer`. We should probably use `CkWallTimer` which actually is documented `Returns the elapsed wall time since the start of execution.`

Reply:
From charm doc 2.2.1.9. Utility Functions
`double CkWallTimer() Returns the elapsed wall time since the start of execution.`

Reply:
Hm, I like that a bit better just because it's partially (time in what units???) documented. I'd be happy to add a commit to this PR to make the switch in `sys::parallel::wall_time()`, if you like. But it might be nice to first confirm the units of `CkWallTimer`...

Reply:
Units is seconds. I'll ask if that can be documented :P